target/riscv/cpu_helper.c

```diff
@@ -622,15 +631,24 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
             /* Fetch access check failed */
             return TRANSLATE_FAIL;
 #if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
-        } else if (access_type == MMU_DATA_CAP_STORE && !(pte & PTE_SC)) {
-            /* SC inhibited */
+        } else if (access_type == MMU_DATA_CAP_STORE &&
+                   !(pte & PTE_SC) && !(pte & PTE_SCM)) {
+            /* SC inhibited, no modifier in place */
             return TRANSLATE_CHERI_FAIL;
 #endif
         } else {
             /* if necessary, set accessed and dirty bits. */
             target_ulong updated_pte = pte | PTE_A |
                 (access_type == MMU_DATA_STORE ? PTE_D : 0);
 
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+            if (access_type == MMU_DATA_CAP_STORE) {
+                /* (SCM and not SC) transitions to (SC and not SCM) */
```

Comment:
So I don't see a check for both bits being set.  If that is reserved, we should be checking that above and returning some other fault instead?  (What fault is defined for invalid settings of PTE bits?)

Similarly for invalid combinations of the load cap bits.

Reply:
Invalid setting of PTE bits is normally the page fault corresponding to the instruction type. I don't know if that means we'd want a page fault or a capability page fault if the CHERI-specific bits were the issue.

Reply:
Hmmm.  x86 has a special code for reserved bits being set (but it also uses a single PF# vector).  I suspect we want to raise the relevant CHERI fault which would happen here by just returning `TRANSLATION_FAIL_CHERI`.

Reply:
I'm torn between raising a CHERI-specific fault or following along with the upstream Sail that maps reserved bit combinations to the PTW_Invalid_PTE internal result (and then on to the generic page fault vector).  See

https://github.com/rems-project/sail-riscv/blob/3026c56e7c2ec5e33a4e9f7e54323b9f8ad6f3aa/model/riscv_pte.sail#L37

https://github.com/rems-project/sail-riscv/blob/3026c56e7c2ec5e33a4e9f7e54323b9f8ad6f3aa/model/riscv_vmem_sv48.sail#L29-31

https://github.com/rems-project/sail-riscv/blob/3026c56e7c2ec5e33a4e9f7e54323b9f8ad6f3aa/model/riscv_ptw.sail#L35

Reply:
I added checks that raise the existing `TRANSLATE_FAIL` code; if anyone feels strongly that this should be `TRANSLATE_CHERI_FAIL`, I'll flip to that.